File: molecular/src/main/java/com/hartwig/actin/molecular/orange/evidence/actionable/VariantEvidence.java
package com.hartwig.actin.molecular.orange.evidence.actionable;

import java.util.List;
import java.util.Set;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.hartwig.actin.molecular.orange.datamodel.purple.PurpleVariant;
import com.hartwig.actin.molecular.orange.evidence.matching.GeneMatching;
import com.hartwig.actin.molecular.orange.evidence.matching.HotspotMatching;
import com.hartwig.actin.molecular.orange.evidence.matching.RangeMatching;
import com.hartwig.serve.datamodel.ActionableEvent;
import com.hartwig.serve.datamodel.ActionableEvents;
import com.hartwig.serve.datamodel.gene.ActionableGene;
import com.hartwig.serve.datamodel.gene.GeneEvent;
import com.hartwig.serve.datamodel.hotspot.ActionableHotspot;
import com.hartwig.serve.datamodel.range.ActionableRange;

import org.jetbrains.annotations.NotNull;

final class VariantEvidence {

    //    this.genes = genes.stream()
    //            .filter(x -> x.event() == GeneLevelEvent.ACTIVATION || x.event() == GeneLevelEvent.INACTIVATION
    //                        || x.event() == GeneLevelEvent.ANY_MUTATION)
    //            .collect(Collectors.toList());
    //
    private static final Set<GeneEvent> APPLICABLE_GENE_EVENTS =
            Sets.newHashSet(GeneEvent.ACTIVATION, GeneEvent.INACTIVATION, GeneEvent.ANY_MUTATION);

    private VariantEvidence() {
    }

    @NotNull
    public static List<ActionableEvent> findMatches(@NotNull ActionableEvents actionableEvents, @NotNull PurpleVariant variant) {
        List<ActionableEvent> applicableEvents = Lists.newArrayList();

        applicableEvents.addAll(hotspotMatches(actionableEvents.hotspots(), variant));
        applicableEvents.addAll(rangeMatches(actionableEvents.ranges(), variant));
        applicableEvents.addAll(geneMatches(actionableEvents.genes(), variant));

        return applicableEvents;
    }

    @NotNull
    private static List<ActionableEvent> hotspotMatches(@NotNull List<ActionableHotspot> hotspots, @NotNull PurpleVariant variant) {
        List<ActionableEvent> matches = Lists.newArrayList();
        for (ActionableHotspot hotspot : hotspots) {
            if (HotspotMatching.isMatch(hotspot, variant)) {
                matches.add(hotspot);
            }
        }
        return matches;
    }

    @NotNull
    private static List<ActionableEvent> rangeMatches(@NotNull List<ActionableRange> ranges, @NotNull PurpleVariant variant) {
        List<ActionableEvent> matches = Lists.newArrayList();
        for (ActionableRange range : ranges) {
            if (RangeMatching.isMatch(range, variant)) {
                matches.add(range);
            }
        }
        return matches;
    }

    @NotNull
    private static List<ActionableEvent> geneMatches(@NotNull List<ActionableGene> genes, @NotNull PurpleVariant variant) {
        List<ActionableEvent> matches = Lists.newArrayList();
        for (ActionableGene gene : genes) {
            if (APPLICABLE_GENE_EVENTS.contains(gene.event()) && GeneMatching.isMatch(gene, variant)) {
                matches.add(gene);
            }
        }
        return matches;
    }
}

File: molecular/src/main/kotlin/com/hartwig/actin/molecular/evidence/actionability/VariantEvidence.kt
package com.hartwig.actin.molecular.evidence.actionability

import com.hartwig.actin.molecular.evidence.matching.GeneMatching
import com.hartwig.actin.molecular.evidence.matching.HotspotMatching
import com.hartwig.actin.molecular.evidence.matching.RangeMatching
import com.hartwig.actin.molecular.evidence.matching.VariantMatchCriteria
import com.hartwig.serve.datamodel.ActionableEvent
import com.hartwig.serve.datamodel.ActionableEvents
import com.hartwig.serve.datamodel.gene.ActionableGene
import com.hartwig.serve.datamodel.gene.GeneEvent
import com.hartwig.serve.datamodel.hotspot.ActionableHotspot
import com.hartwig.serve.datamodel.range.ActionableRange

internal class VariantEvidence private constructor(
    private val actionableHotspots: List<ActionableHotspot>,
    private val actionableRanges: List<ActionableRange>, private val applicableActionableGenes: List<ActionableGene>
) : EvidenceMatcher<VariantMatchCriteria> {

    override fun findMatches(event: VariantMatchCriteria): List<ActionableEvent> {
        return listOf(hotspotMatches(event), rangeMatches(event), geneMatches(event)).flatten()
    }

    private fun hotspotMatches(variant: VariantMatchCriteria): List<ActionableEvent> {
        return filterMatchingEvents(actionableHotspots, variant, HotspotMatching::isMatch)
    }

    private fun rangeMatches(variant: VariantMatchCriteria): List<ActionableEvent> {
        return filterMatchingEvents(actionableRanges, variant, RangeMatching::isMatch)
    }

    private fun geneMatches(variant: VariantMatchCriteria): List<ActionableEvent> {
        return filterMatchingEvents(applicableActionableGenes, variant, GeneMatching::isMatch)
    }

    private fun <T : ActionableEvent> filterMatchingEvents(
        events: List<T>, variant: VariantMatchCriteria, predicate: (T, VariantMatchCriteria) -> Boolean
    ): List<ActionableEvent> {
        return if (!variant.isReportable) emptyList() else events.filter { predicate.invoke(it, variant) }
    }

    companion object {
        private val APPLICABLE_GENE_EVENTS = setOf(GeneEvent.ACTIVATION, GeneEvent.INACTIVATION, GeneEvent.ANY_MUTATION)

        fun create(actionableEvents: ActionableEvents): VariantEvidence {
            val applicableActionableGenes = actionableEvents.genes()
                .filter { actionableGene: ActionableGene -> APPLICABLE_GENE_EVENTS.contains(actionableGene.event()) }
            val ranges = listOf(actionableEvents.codons(), actionableEvents.exons()).flatten()
            return VariantEvidence(actionableEvents.hotspots(), ranges, applicableActionableGenes)
        }
    }
}